File: frontend/src/api/daily-intake.ts
import api from './client'
import type {
  DailyIntakeCategory,
  DailyIntakeAsrProviderSelection,
  DailyIntakeItem,
  DailyIntakeParseResponse,
  DailyIntakeSheet,
  DailyIntakeSheetSummary,
  DailyIntakeSource,
} from '../features/daily-intake/types'

export interface DailyIntakeSheetResponse {
  success: boolean
  message: string
  sheet: DailyIntakeSheet
}

export interface DailyIntakeHistoryResponse {
  success: boolean
  message: string
  sheets: DailyIntakeSheetSummary[]
}

export interface DailyIntakeItemMutationResponse {
  success: boolean
  message: string
  item: DailyIntakeItem | null
  sheet: DailyIntakeSheet
  merged: boolean
}

export interface DailyIntakeDeleteResponse {
  success: boolean
  message: string
  sheet: DailyIntakeSheet
}

export interface DailyIntakeSpeechCapabilitiesResponse {
  success: boolean
  stable_transcription_enabled: boolean
  provider: string | null
  model: string | null
  requested_device?: string | null
  requested_compute_type?: string | null
  device?: string | null
  compute_type?: string | null
  fallback_used?: boolean
  fallback_reason?: string | null
  primary_provider?: string | null
  backup_provider?: string | null
  failover_enabled?: boolean
  shadow_compare_enabled?: boolean
  providers?: Array<Record<string, unknown>>
  message: string
}

export interface DailyIntakeSpeechRuntimeDiagnosticsResponse {
  success: boolean
  dependency_available: boolean
  provider?: string | null
  model?: string | null
  requested_device?: string | null
  requested_compute_type?: string | null
  resolved_device?: string | null
  resolved_compute_type?: string | null
  effective_device?: string | null
  effective_compute_type?: string | null
  cuda_device_count: number
  supported_compute_types_cpu: string[]
  supported_compute_types_cuda: string[]
  missing_cuda_runtime_dlls: string[]
  model_loaded: boolean
  runtime_checked: boolean
  fallback_used: boolean
  fallback_reason?: string | null
  suggested_fix?: string | null
  primary_provider?: string | null
  backup_provider?: string | null
  failover_enabled?: boolean
  shadow_compare_enabled?: boolean
  providers?: Array<Record<string, unknown>>
  message: string
}

export function getTodayDailyIntake() {
  return api.get<DailyIntakeSheetResponse>('/api/daily-intake/today')
}

export function getDailyIntakeByDate(intakeDate: string) {
  return api.get<DailyIntakeSheetResponse>(`/api/daily-intake/${encodeURIComponent(intakeDate)}`)
}

export function getDailyIntakeHistory(limit = 30) {
  return api.get<DailyIntakeHistoryResponse>('/api/daily-intake/history', {
    params: { limit },
  })
}

export function getDailyIntakeSpeechCapabilities() {
  return api.get<DailyIntakeSpeechCapabilitiesResponse>('/api/daily-intake/speech-capabilities')
}

export function getDailyIntakeSpeechRuntimeDiagnostics() {
  return api.get<DailyIntakeSpeechRuntimeDiagnosticsResponse>('/api/daily-intake/speech-runtime-diagnostics')
}

export function createDailyIntakeItem(params: {
  intake_date: string
  name: string
  category: DailyIntakeCategory
  quantity: number
  unit: string
  source?: DailyIntakeSource
  transcript?: string
}) {
  return api.post<DailyIntakeItemMutationResponse>('/api/daily-intake/items', params)
}

export function updateDailyIntakeItem(
  itemId: number,
  params: {
    name: string
    category: DailyIntakeCategory
    quantity: number
    unit: string
    source?: DailyIntakeSource | null
    transcript?: string | null
  },
) {
  return api.put<DailyIntakeItemMutationResponse>(`/api/daily-intake/items/${itemId}`, params)
}

export function deleteDailyIntakeItem(itemId: number) {
  return api.delete<DailyIntakeDeleteResponse>(`/api/daily-intake/items/${itemId}`)
}

export function transcribeDailyIntakeAudio(params: {
  intake_date: string
  audio: Blob
  filename?: string
  category?: DailyIntakeCategory
  asr_provider?: DailyIntakeAsrProviderSelection
  fallback_enabled?: boolean
}) {
  const formData = new FormData()
  formData.append('intake_date', params.intake_date)
  if (params.category) {
    formData.append('category', params.category)
  }
  if (params.asr_provider) {
    formData.append('asr_provider', params.asr_provider)
  }
  if (typeof params.fallback_enabled === 'boolean') {
    formData.append('fallback_enabled', String(params.fallback_enabled))
  }
  formData.append('audio', params.audio, params.filename || 'daily-intake-recording.webm')

  return api.post<DailyIntakeParseResponse>('/api/daily-intake/transcribe-audio', formData, {
    timeout: 180000,
  })
}

File: backend/api/routes/daily_intake.py
from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile
from fastapi.responses import FileResponse
from starlette.concurrency import run_in_threadpool

from backend.auth.dependencies import require_permission
from backend.models.schemas import (
    DailyIntakeDeleteResponse,
    DailyIntakeHistoryResponse,
    DailyIntakeItemMutationResponse,
    DailyIntakeItemUpdateRequest,
    DailyIntakeItemUpsertRequest,
    DailyIntakeParseRequest,
    DailyIntakeParseResponse,
    DailyIntakeSheetResponse,
    DailyIntakeSpeechCapabilitiesResponse,
    DailyIntakeSpeechDiagnosticsResponse,
)
from backend.diagnostics.asr_self_check import run_asr_self_check
from backend.services.daily_intake_asr_service import DailyIntakeAsrError, DailyIntakeAsrService
from backend.services.daily_intake_service import DailyIntakeService

router = APIRouter()
service = DailyIntakeService()
speech_to_text_service = DailyIntakeAsrService(daily_intake_service=service)


def _raise_http_error(exc: Exception) -> None:
    if isinstance(exc, KeyError):
        raise HTTPException(status_code=404, detail=exc.args[0]) from exc
    raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/today", response_model=DailyIntakeSheetResponse, dependencies=[Depends(require_permission("daily_check:view"))])
def get_today_daily_intake_sheet():
    return DailyIntakeSheetResponse(**service.get_today_sheet())


@router.get("/history", response_model=DailyIntakeHistoryResponse, dependencies=[Depends(require_permission("daily_check:view"))])
def list_daily_intake_history(limit: int = Query(default=30, ge=1, le=365)):
    return DailyIntakeHistoryResponse(**service.list_history(limit=limit))


@router.get(
    "/speech-capabilities",
    response_model=DailyIntakeSpeechCapabilitiesResponse,
    dependencies=[Depends(require_permission("daily_check:view"))],
)
def get_daily_intake_speech_capabilities():
    return DailyIntakeSpeechCapabilitiesResponse(**speech_to_text_service.capabilities())


@router.get(
    "/asr-self-check",
    dependencies=[Depends(require_permission("daily_check:view"))],
)
def get_daily_intake_asr_self_check():
    """Dump loaded ASR module locations for diagnosing stub fallbacks."""
    return run_asr_self_check(strict=False)


@router.get(
    "/speech-runtime-diagnostics",
    response_model=DailyIntakeSpeechDiagnosticsResponse,
    dependencies=[Depends(require_permission("daily_check:view"))],
)
def get_daily_intake_speech_runtime_diagnostics(
    probe_runtime: bool = Query(default=True),
):
    return DailyIntakeSpeechDiagnosticsResponse(
        **speech_to_text_service.diagnostics(probe_runtime=probe_runtime)
    )


@router.get("/asr-shadow-compare", dependencies=[Depends(require_permission("daily_check:view"))])
def list_daily_intake_asr_shadow_compare(limit: int = Query(default=50, ge=1, le=500)):
    records = speech_to_text_service.shadow_store.read_recent(limit=limit)
    return {
        "success": True,
        "path": str(speech_to_text_service.shadow_store.path),
        "total_returned": len(records),
        "records": records,
    }


@router.get("/asr-shadow-compare/export", dependencies=[Depends(require_permission("daily_check:export"))])
def export_daily_intake_asr_shadow_compare():
    path = speech_to_text_service.shadow_store.export_path()
    return FileResponse(
        path,
        media_type="application/x-ndjson",
        filename=path.name,
    )


@router.get("/{intake_date}", response_model=DailyIntakeSheetResponse, dependencies=[Depends(require_permission("daily_check:view"))])
def get_daily_intake_sheet(intake_date: str):
    try:
        return DailyIntakeSheetResponse(**service.get_sheet(intake_date))
    except (ValueError, KeyError) as exc:
        _raise_http_error(exc)


@router.post("/items", response_model=DailyIntakeItemMutationResponse, dependencies=[Depends(require_permission("daily_check:create"))])
def create_daily_intake_item(req: DailyIntakeItemUpsertRequest):
    try:
        result = service.add_item(
            intake_date=req.intake_date,
            name=req.name,
            category=req.category,
            quantity=req.quantity,
            unit=req.unit,
            source=req.source,
            transcript=req.transcript,
        )
        return DailyIntakeItemMutationResponse(**result)
    except (ValueError, KeyError) as exc:
        _raise_http_error(exc)


@router.put(
    "/items/{item_id}",
    response_model=DailyIntakeItemMutationResponse,
    dependencies=[Depends(require_permission("daily_check:update"))],
)
def update_daily_intake_item(item_id: int, req: DailyIntakeItemUpdateRequest):
    try:
        result = service.update_item(
            item_id,
            name=req.name,
            category=req.category,
            quantity=req.quantity,
            unit=req.unit,
            source=req.source,
            transcript=req.transcript,
        )
        return DailyIntakeItemMutationResponse(**result)
    except (ValueError, KeyError) as exc:
        _raise_http_error(exc)


@router.delete(
    "/items/{item_id}",
    response_model=DailyIntakeDeleteResponse,
    dependencies=[Depends(require_permission("daily_check:delete"))],
)
def delete_daily_intake_item(item_id: int):
    try:
        return DailyIntakeDeleteResponse(**service.delete_item(item_id))
    except (ValueError, KeyError) as exc:
        _raise_http_error(exc)


@router.post(
    "/parse-transcript",
    response_model=DailyIntakeParseResponse,
    dependencies=[Depends(require_permission("daily_check:create"))],
)
def parse_daily_intake_transcript(req: DailyIntakeParseRequest):
    try:
        return DailyIntakeParseResponse(
            **service.parse_transcript(
                transcript=req.transcript,
                intake_date=req.intake_date,
                category=req.category,
            )
        )
    except ValueError as exc:
        _raise_http_error(exc)


@router.post(
    "/transcribe-audio",
    response_model=DailyIntakeParseResponse,
    dependencies=[Depends(require_permission("daily_check:create"))],
)
async def transcribe_daily_intake_audio(
    intake_date: str = Form(...),
    category: str | None = Form(default=None),
    asr_provider: str | None = Form(default="auto"),
    fallback_enabled: str | None = Form(default=None),
    audio: UploadFile = File(...),
):
    try:
        payload = await run_in_threadpool(
            speech_to_text_service.transcribe_audio,
            file_bytes=await audio.read(),
            filename=audio.filename or "daily-intake-recording.webm",
            content_type=audio.content_type,
            intake_date=intake_date,
            category=category,
            asr_provider=asr_provider,
            fallback_enabled=_parse_optional_bool(fallback_enabled),
        )
        return DailyIntakeParseResponse(**payload)
    except DailyIntakeAsrError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=502, detail=str(exc)) from exc
    except ValueError as exc:
        _raise_http_error(exc)


def _parse_optional_bool(raw_value: str | None) -> bool | None:
    if raw_value is None:
        return None
    normalized = str(raw_value).strip().lower()
    if not normalized:
        return None
    return normalized not in {"0", "false", "no"}
